src/docusign-po-failure/test/logic.js

```diff
@@ -86,123 +91,187 @@ describe('Logic', () => {
+                        name : 'actualPrice',
+                        value: '2000'
+                    },
+                    {
+                        $class: 'com.docusign.connect.CustomField',
+                        name : 'currencyCode',
+                        value: 'USD'
+                    }
+                ],
+                timestamp: '2019-03-10T17:38:01.412Z'
+            };
+            const state = {};
+            state.$class = 'com.docusign.clauses.PurchaseOrderFailureState';
+            state.stateId = 'org.accordproject.cicero.contract.AccordContractState#1';
+            state.pastFailures = [];
+            state.nbPastFailures = 0;
+            const result = await engine.execute(clause, request, state);
+            result.should.not.be.null;
+            result.response.penalty.should.deep.equal({
+                '$class': 'org.accordproject.money.MonetaryAmount',
+                'doubleValue': 200,
+                'currencyCode': 'USD'
+            });
+            result.state.nbPastFailures.should.equal(1);
+            result.emit[0].$class.should.equal('org.accordproject.cicero.runtime.PaymentObligation');
+            result.emit[0].amount.should.deep.equal({
+                '$class': 'org.accordproject.money.MonetaryAmount',
+                'doubleValue': 200,
+                'currencyCode': 'USD'
+            });
+        });
+        
+        it('should calculate the penalty when delivered after first late period', async function () {
+            const request = {
+                $class: 'com.docusign.connect.DocuSignEnvelopeInformation',
+                envelopeStatus: {
+                    $class: 'com.docusign.connect.EnvelopeStatus',
+                    status : 'Completed'
+                },
+                customFields: [
+                    {
+                        $class: 'com.docusign.connect.CustomField',
+                        name : 'deliveryDate',
+                        value: '2019-03-10'
+                    },
+                    {
+                        $class: 'com.docusign.connect.CustomField',
+                        name : 'actualPrice',
+                        value: '2000'
+                    },
+                    {
+                        $class: 'com.docusign.connect.CustomField',
+                        name : 'currencyCode',
+                        value: 'USD'
+                    }
+                ],
+                timestamp: '2019-03-10T17:38:01.412Z'
+            };
+            const state = {};
+            state.$class = 'com.docusign.clauses.PurchaseOrderFailureState';
+            state.stateId = 'org.accordproject.cicero.contract.AccordContractState#1';
+            state.pastFailures = [];
+            state.nbPastFailures = 0;
+            const result = await engine.execute(clause, request, state);
+            result.should.not.be.null;
+            result.response.penalty.should.deep.equal({
```

Comment:
If the delivery date is 3-10 and the time stamp is 3-10, shouldn’t this test case not result in a penalty?

Reply:
The timestamp is not taken into account by the Cicero/Ergo execution. Only real current time `Moment.now()` really. That's why the tests are written this way. Note that the tests will start failing soon!

All of that is fixed in development for Cicero and Ergo -- The tests should be rewritten with Cucumber which lets you set a current time explicitly as part of the test.

Reply:
Ah, yes, you did mention this already, sorry!